pest/src/stack.rs

```diff
@@ -0,0 +1,177 @@
+// pest. The Elegant Parser
+// Copyright (c) 2018 Dragoș Tiselice
+//
+// Licensed under the Apache License, Version 2.0
+// <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0> or the MIT
+// license <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. All files in the project carrying such notice may not be copied,
+// modified, or distributed except according to those terms.
+
+/// Implementation of a `Stack` which maintains an log of `StackOp`s in order to rewind the stack
+/// to a previous state.
+#[derive(Debug)]
+pub struct Stack<T: Clone> {
+    ops: Vec<StackOp<T>>,
+    cache: Vec<T>,
+    snapshots: Vec<usize>
+}
+
+impl<T: Clone> Stack<T> {
+    /// Creates a new `Stack`.
+    pub fn new() -> Self {
+        Stack {
+            ops: vec![],
+            cache: vec![],
+            snapshots: vec![]
+        }
+    }
+
+    /// Returns `true` if the stack is currently empty.
+    pub fn is_empty(&self) -> bool {
+        self.cache.is_empty()
+    }
+
+    /// Returns the top-most `&T` in the `Stack`.
+    pub fn peek(&self) -> Option<&T> {
+        self.cache.last()
+    }
+
+    /// Pushes a `T` onto the `Stack`.
+    pub fn push(&mut self, elem: T) {
+        self.ops.push(StackOp::Push(elem.clone()));
+        self.cache.push(elem);
+    }
+
+    /// Pops the top-most `T` from the `Stack`.
+    pub fn pop(&mut self) -> Option<T> {
+        self.ops.push(StackOp::Pop);
+        self.cache.pop()
+    }
+
+    /// Clears all the values from the `Stack`.
+    pub fn clear(&mut self) {
+        self.ops.push(StackOp::Clear);
+        self.cache.clear();
+    }
+
+    /// Takes a snapshot of the current `Stack`.
+    pub fn snapshot(&mut self) {
+        self.snapshots.push(self.ops.len());
+    }
+
+    /// Rewinds the `Stack` to the most `snapshot()`. If no `snapshot()` has been taken, this
+    /// function will do nothing.
+    pub fn backtrack(&mut self) {
+        match self.snapshots.pop() {
+            Some(index) => {
+                self.ops.truncate(index);
+                self.recache();
+            }
+            None => {}
+        }
+    }
+
+    // Rebuilds the internal cache based on the previous operations stored.
```

Comment:
It might make sense to have an implementation that takes better advantage of the cache by not clearing it, instead making snapshots of the stack content every now and then. (when I say content, I mean top and concatenated content, in which case a less general `&str` instead of `T` implementation would make sense)

Reply:
Not quite sure I follow there. When you keep popping off values from the stack, you need to know where the boundaries of the individual strings are no? A top and concatenated properties would lose context no?